core/capabilities/webapi/outgoing_connector_handler.go

```diff
@@ -95,55 +103,85 @@ func (c *OutgoingConnectorHandler) HandleSingleNodeRequest(ctx context.Context,
 	ctx, cancel := context.WithTimeout(ctx, timeoutDuration+margin)
 	defer cancel()
 
-	payload, err := json.Marshal(req)
-	if err != nil {
-		return nil, fmt.Errorf("failed to marshal fetch request: %w", err)
+	operation := c.toRetryableFunc(ctx, messageID, req)
+
+	if req.WithRetries == nil || !*req.WithRetries {
+		return operation()
 	}
 
-	ch, err := c.responses.new(messageID)
-	if err != nil {
-		return nil, fmt.Errorf("duplicate message received for ID: %s", messageID)
+	maxTries := uint(defaultMaxTries)
+	if req.MaxTries != nil && *req.MaxTries <= defaultMaxTries {
+		maxTries = *req.MaxTries
 	}
-	defer c.responses.cleanup(messageID)
 
-	lggr.Debugw("sending request to gateway")
+	return backoff.Retry(ctx, operation, c.getBackoffOpts(time.Duration(req.TimeoutMs)*time.Millisecond, maxTries)...)
+}
 
-	body := &api.MessageBody{
-		MessageId: messageID,
-		DonId:     c.gc.DonID(),
-		Method:    c.method,
-		Payload:   payload,
+func (c *OutgoingConnectorHandler) getBackoffOpts(d time.Duration, maxTries uint) []backoff.RetryOption {
+	if c.toBackoffOpts != nil {
+		return c.toBackoffOpts(d, maxTries)
 	}
 
-	selectedGateway, err := c.awaitConnection(ctx, awaitContext{
-		messageID:  messageID,
-		workflowID: req.WorkflowID,
-	})
-	if err != nil {
-		return nil, err
+	return []backoff.RetryOption{
+		backoff.WithBackOff(backoff.NewExponentialBackOff()),
+		backoff.WithMaxElapsedTime(d),
+		backoff.WithMaxTries(maxTries),
 	}
+}
 
-	if err := c.gc.SignAndSendToGateway(ctx, selectedGateway, body); err != nil {
-		return nil, errors.Wrap(err, "failed to send request to gateway")
-	}
+func (c *OutgoingConnectorHandler) toRetryable(ctx context.Context, messageID string, req capabilities.Request) func() (*api.Message, error) {
```

Comment:
I find this function name to be confusing.

The logic of this function isn't purely wrapping another function as retryable. It is actually doing a lot more than that. With this current name there is some at a glance context lost on what the other 50 lines of code are doing.

I'd also recommend some function signature doc on why we need "toRetryable". (I see in the struct it has what it is, but some words on why would help for grokking)

Reply:
^ Justin is making a good point here -- this function actual contains the heart of your logic; it should be called sendRequest or something similar

Reply:
point taken, but this is really just a refactor.  I'm extracting all the existing logic and wrapping it to close over the context, message id and request so it can be retried.